src/librustc/infer/error_reporting/need_type_info.rs

```diff
@@ -208,6 +215,20 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             // 3 |     let _ = x.sum() as f64;
             //   |               ^^^ cannot infer type for `S`
             span
+        } else if let Some(ExprKind::MethodCall(_, call_span, _)) = local_visitor.found_method_call {
```

Comment:
So I think we really ought to check that:

* the method expects exactly 1 type parameter
* the declared return type of the method is equal to this type parameter